pysteps/nowcasts/steps.py

```diff
@@ -33,6 +36,971 @@
+    noise_stddev_adj: Optional[str] = None
+    ar_order: int = 2
+    velocity_perturbation_method: Optional[str] = "bps"
+    conditional: bool = False
+    probmatching_method: Optional[str] = "cdf"
+    mask_method: Optional[str] = "incremental"
+    seed: Optional[int] = None
+    num_workers: int = 1
+    fft_method: str = "numpy"
+    domain: str = "spatial"
+    extrapolation_kwargs: Dict[str, Any] = field(default_factory=dict)
+    filter_kwargs: Dict[str, Any] = field(default_factory=dict)
+    noise_kwargs: Dict[str, Any] = field(default_factory=dict)
+    velocity_perturbation_kwargs: Dict[str, Any] = field(default_factory=dict)
+    mask_kwargs: Dict[str, Any] = field(default_factory=dict)
+    measure_time: bool = False
+    callback: Optional[Callable[[Any], None]] = None
+    return_output: bool = True
+
+
+@dataclass
+class StepsNowcasterParams:
+    fft: Any = None
+    bandpass_filter: Any = None
+    extrapolation_method: Any = None
+    decomposition_method: Any = None
+    recomposition_method: Any = None
+    noise_generator: Optional[callable] = None
+    perturbation_generator: Optional[callable] = None
+    noise_std_coefficients: Optional[np.ndarray] = None
+    ar_model_coefficients: Optional[np.ndarray] = None  # Corresponds to phi
+    autocorrelation_coefficients: Optional[np.ndarray] = None  # Corresponds to gamma
+    domain_mask: Optional[np.ndarray] = None
+    structuring_element: Optional[np.ndarray] = None
+    precipitation_mean: Optional[float] = None
+    wet_area_ratio: Optional[float] = None
+    mask_rim: Optional[int] = None
+    num_ensemble_workers: int = 1
+    xy_coordinates: Optional[np.ndarray] = None
+    velocity_perturbation_parallel: Optional[List[float]] = None
+    velocity_perturbation_perpendicular: Optional[List[float]] = None
+
+
+@dataclass
+class StepsNowcasterState:
+    precip_forecast: Optional[List[Any]] = field(default_factory=list)
+    precip_cascades: Optional[List[List[np.ndarray]]] = field(default_factory=list)
+    precip_decomposed: Optional[List[Dict[str, Any]]] = field(default_factory=list)
+    # The observation mask (where the radar can observe the precipitation)
+    precip_mask: Optional[List[Any]] = field(default_factory=list)
+    precip_mask_decomposed: Optional[Dict[str, Any]] = field(default_factory=dict)
+    # The mask around the precipitation fields (to get only non-zero values)
+    mask_precip: Optional[np.ndarray] = None
+    mask_threshold: Optional[np.ndarray] = None
+    random_generator_precip: Optional[List[np.random.RandomState]] = field(
+        default_factory=list
+    )
+    random_generator_motion: Optional[List[np.random.RandomState]] = field(
+        default_factory=list
+    )
+    velocity_perturbations: Optional[List[callable]] = field(default_factory=list)
+    fft_objects: Optional[List[Any]] = field(default_factory=list)
+
+
+class StepsNowcaster:
+    def __init__(self, precip, velocity, time_steps, steps_config):
```

Comment:
The type hints in the dataclasses already help a lot. As an extra bonus these classes can then also be used as type hints for the methods that expect them.

Also, I believe this `__init__` method (or the class itself) should have a docstring indicating to the user what they are expected to provide as arguments, or is that done somewhere else currently?

```suggestion
    def __init__(self, precip, velocity, time_steps, steps_config: StepsNowcasterConfig):
```

Reply:
I think that the second part of this comment maybe still needs to be addressed? (part of) the docstring from the `forecast` method probably needs to be copied here so that when users use the class directly they know what to do with each parameter. Possibly the docstring can be cut into 2 pieces where the part that describes `precip`, `velocity` and `timesteps` can be moved here and the part that describes all other parameters could be a docstring for the `StepsNowcasterConfig` dataclass.

Reply:
This should be done now